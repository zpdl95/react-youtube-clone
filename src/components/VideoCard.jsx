import React from 'react';
import { useNavigate } from 'react-router-dom';
import { formatAgo } from '../util/date';

export default function VideoCard({ video, type }) {
  const { publishedAt, title, thumbnails, channelTitle } = video.snippet;
  const isList = type === 'list';
  const navigate = useNavigate();

  const handleClick = () => {
    navigate(`/videos/watch/${video.id}`, {
      state: { video },
    });
  };

  return (
    <li className={`flex ${isList ? 'gap-2' : 'flex-col mb-8'} `}>
      <img
        className={`cursor-pointer ${
          isList ? 'w-40 rounded-md' : 'rounded-lg'
        }`}
        onClick={handleClick}
        src={thumbnails.medium.url}
        alt={title}
      />

      <div>
        <p
          className={`font-semibold my-2 line-clamp-2 ${
            isList ? 'text-sm font-normal' : ''
          }`}
        >
          {title}
        </p>
        <p className={`text-sm opacity-70 ${isList ? 'text-xs' : ''}`}>
          {channelTitle}
        </p>
        <p className={`text-sm opacity-70 ${isList ? 'text-xs' : ''}`}>
          {formatAgo(publishedAt)}
        </p>
      </div>
    </li>
  );
}
