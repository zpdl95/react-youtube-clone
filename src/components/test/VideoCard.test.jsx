import React from 'react';
import { render, screen } from '@testing-library/react';
import { Route, useLocation } from 'react-router-dom';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';
import VideoCard from '../VideoCard';
import { formatAgo } from '../../util/date';
import { fakeVideo as video } from '../../test/videos';
import { withRouter } from '../../test/utils';

describe('VideoCard', () => {
  const { title, channelTitle, publishedAt, thumbnails } = video.snippet;

  it('renders grid type correctly', () => {
    const component = renderer.create(
      withRouter(<Route path='/' element={<VideoCard video={video} />} />)
    );

    // toMatchSnapshot() 함수가 처음실행하면 스냅샷을 만들고 두번째 실행부터 test를 한다
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('renders list type correctly', () => {
    const component = renderer.create(
      withRouter(
        <Route path='/' element={<VideoCard video={video} type='list' />} />
      )
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('renders video item', () => {
    render(
      withRouter(<Route path='/' element={<VideoCard video={video} />} />)
    );

    const image = screen.getByRole('img');

    expect(image.src).toBe(thumbnails.medium.url);
    expect(image.alt).toBe(title);
    expect(screen.getByText(title)).toBeInTheDocument();
    expect(screen.getByText(channelTitle)).toBeInTheDocument();
    expect(screen.getByText(formatAgo(publishedAt))).toBeInTheDocument();
  });

  it('navigates to detailed video page with video state when clicked', async () => {
    function LocationStateDisplay() {
      return <pre>{JSON.stringify(useLocation().state)}</pre>;
    }
    render(
      withRouter(
        <>
          <Route path='/' element={<VideoCard video={video} />} />
          <Route
            path={`/videos/watch/${video.id}`}
            element={<LocationStateDisplay />}
          />
        </>
      )
    );

    const img = screen.getByRole('img');
    await userEvent.click(img);

    expect(screen.getByText(JSON.stringify({ video }))).toBeInTheDocument();
  });
});
