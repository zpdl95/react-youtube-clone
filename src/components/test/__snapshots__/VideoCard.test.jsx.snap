// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VideoCard > renders grid type correctly 1`] = `
<li
  className="flex flex-col mb-8 "
>
  <img
    alt="title"
    className="cursor-pointer rounded-lg"
    onClick={[Function]}
    src="http://image/"
  />
  <div>
    <p
      className="font-semibold my-2 line-clamp-2 "
    >
      title
    </p>
    <p
      className="text-sm opacity-70 "
    >
      channelTitle
    </p>
    <p
      className="text-sm opacity-70 "
    >
      just now
    </p>
  </div>
</li>
`;

exports[`VideoCard > renders list type correctly 1`] = `
<li
  className="flex gap-2 "
>
  <img
    alt="title"
    className="cursor-pointer w-40 rounded-md"
    onClick={[Function]}
    src="http://image/"
  />
  <div>
    <p
      className="font-semibold my-2 line-clamp-2 text-sm font-normal"
    >
      title
    </p>
    <p
      className="text-sm opacity-70 text-xs"
    >
      channelTitle
    </p>
    <p
      className="text-sm opacity-70 text-xs"
    >
      just now
    </p>
  </div>
</li>
`;
